### PR TITLE
fix no-duplicate-super validates with ternary operator

### PR DESCRIPTION
**Pull request by mateuszwitkowski**
#### PR checklist

- [x ] Addresses an existing issue: #3275
- [x ] New feature, bugfix, or enhancement
  - [x ] Includes tests
- [ ] Documentation update

#### Overview of change:
Added case for ternary operator in no-duplicate-super rule.

#### CHANGELOG.md entry:

[bugfix] using ternary operator for calling super() now passes no-duplicate-super rule.

### CODE REVIEW

File: src/rules/noDuplicateSuperRule.ts [hunk 1]
```diff
@@ -80,6 +80,11 @@ function walk(ctx: Lint.WalkContext<void>): void {
                     ? { node: node.parent! as ts.CallExpression, break: false }
                     : Kind.NoSuper;
 
+            case ts.SyntaxKind.ConditionalExpression: {
+                const { whenTrue, whenFalse } = node as ts.ConditionalExpression;
+                return worse(getSuperForNode(whenTrue), whenFalse !== undefined ? getSuperForNode(whenFalse) : Kind.NoSuper);
```
Comment (ajafff) [inline src/rules/noDuplicateSuperRule.ts]: `whenFalse` can't be undefined

Comment (mateuszwitkowski) [reply src/rules/noDuplicateSuperRule.ts]: You're right of course. I'll fix it. There's another issue though. If someone would use nested ternary operators super() will be called twice:

```class Ternary extends Base {
	constructor(props?: any) {
        props ? super(props) : super() ? super() : super(); // if no props passed super() is called twice
	}
}
```
Do you think such a case should be covered? If so can you give me a tip how to handle it?

File: src/rules/noDuplicateSuperRule.ts [hunk 2]
```diff
@@ -80,6 +80,11 @@ function walk(ctx: Lint.WalkContext<void>): void {
                     ? { node: node.parent! as ts.CallExpression, break: false }
                     : Kind.NoSuper;
 
+            case ts.SyntaxKind.ConditionalExpression: {
+                const { whenTrue, whenFalse } = node as ts.ConditionalExpression;
+                return worse(getSuperForNode(whenTrue), getSuperForNode(whenFalse));
```
Comment (ajafff) [inline src/rules/noDuplicateSuperRule.ts]: to also handle `super()` calls inside the condition:
```ts
const inCondition = getSuperForNode(node.condition);
const inBranches = worse(getSuperForNode(whenTrue), getSuperForNode(whenFalse));
if (typeof inCondition !== "number" && typeof inBranches !== number) {
    addDuplicateFailure(inCondition.node, inBranches.node);
}
return worse(inCondition, inBranches);
```

*code is untested and written only in GitHub comment editor.

File: test/rules/no-duplicate-super/test.ts.lint [hunk 3]
```diff
@@ -259,5 +259,14 @@ declare const n: number;
     }
 }
 
+// With ternary operator
+{
+    class {
```
Comment (ajafff) [inline test/rules/no-duplicate-super/test.ts.lint]: btw. this is not a valid class declaration. Class declarations need a name.

Comment (mateuszwitkowski) [reply test/rules/no-duplicate-super/test.ts.lint]: I thought about it but that's how it's done in all the other test cases for that rule. I've figured out there's some reason for it and followed it. I'll fix it in all test cases with updated PR.

**Comment by ajafff** [comment]
Thanks @mateuszwitkowski

**Pull request merged**